### ISSUE TITLE
Accessing `user_data` when building a `ViewWrapper`?

### ISSUE THREAD

**Issue by yaymukund** [gyscos/Cursive#330]
Question

# Problem description

TL;DR: It seems like `user_data` is mainly accessible when handling callbacks, but I'd like to render a deeply-nested `TextView` with some `user_data()` property.

Presently, I'm doing this:

```rust
impl ViewWrapper for PlayerView {
    wrap_impl!(self.view: LinearLayout);
}

impl PlayerView {
    pub fn new(elapsed: TextContent) -> PlayerView {
        let view = LinearLayout::horizontal()
            .child(TextView::new_with_content(elapsed));

        PlayerView { view }
    }
}
```
But this means I have to pass my `TextContent` manually down the different layers of my app. It would be really nice to do something like:

```rust
let siv = Cursive::new();
siv.set_user_data(elapsed);

// this can be very deeply nested
impl PlayerView {
    wrap_impl!(view: LayoutView);

    fn wrap_build(app: &Cursive) {
        let elapsed = TextView::new_with_content(app.user_data().elapsed);
        // build rest of view...
    }
}
```

Obviously, that isn't possible because `wrap_build()` doesn't exist! Is there a good way to get a hold of `user_data()` many levels deep into a Cursive app without passing it down each level?

Thank you for building cursive - it's been a treat to use.

# Environment

* Cursive version: 0.11.1 from crates.io

**Comment by gyscos** [comment]
Hi, and thanks for the report!

I'm not exactly sure what you want to achieve from the example - do you need the user data when creating the view? Usually, views are created when a `&mut Cursive` handle is available; when exactly are you planning on creating this view?

That being said, I realize there are more places when this user data would be needed, namely during `View::print`, `View::layout` and `View::required_size` (and maybe other places as well).

For `View::print`, a solution is to embed a reference to the user data in the `Printer`.
For `View::layout` and `View::required_size`, it's a bit more complicated...
For a while, I wanted to include some context to these methods, to indicate things like the view position in the graph, and possibly a way to get the user data.

The risk is that adding some context like this to every method of the `View` trait may end up adding quite a bit of complexity for little gain for most people, so I'm still balancing the options...

**Comment by yaymukund** [comment]
Thank you for the detailed reply! I'll have a closer look and get back to you tonight.

**Comment by yaymukund** [comment]
Whew, sorry for the delayed response. 

You are correct that the `&mut Cursive` handle is available at the top level. However, my concern is that in composing a hierarchy of views, you have to pass down the state from layer to layer. Consider this simplified example:

```rust
fn main() {
    let player =  Player::new();
    let app = ApplicationView::new(&player);
}

impl ApplicationView {
    pub fn new(player: &Player) {
        let nav = NavView::new(player);
    }
}

impl NavView {
    pub fn new(player: &Player) {
        let player = PlayerView::new(player);
        // ...
  }
}

impl PlayerView {
    pub fn new(player: &Player) {
        // the only place we use the player directly
    }
}
```

I agree adding some sort of global state management might be more trouble than it's worth but I just want to make sure I'm not missing some more obvious pattern.

**Comment by yaymukund** [comment]
i have just been passing things down and it's fine
